packages/safe-ds-lang/src/language/generation/safe-ds-python-generator.ts

```diff
@@ -381,20 +395,238 @@ export class SafeDsPythonGenerator {
+        // Find dependencies matching on impurity information
+        const dependencies = new Set<SdsStatement>();
+        for (const statement of statementsWithEffect) {
+            // Only statements before the current assignment can affect the current assignment
+            if (statement.$containerIndex! < subject.$containerIndex!) {
+                let pastImpurityReasons = [];
+                if (isSdsAssignment(statement)) {
+                    pastImpurityReasons = this.purityComputer.getImpurityReasonsForExpression(statement.expression);
+                } else if (isSdsExpressionStatement(statement)) {
+                    pastImpurityReasons = this.purityComputer.getImpurityReasonsForExpression(statement.expression);
+                } else {
+                    /* c8 ignore next 2 */
+                    continue;
+                }
+                for (const futureImpurityReason of impurityReasons) {
+                    for (const pastImpurityReason of pastImpurityReasons) {
+                        if (this.hasImpurityReasonEffectOnFuture(pastImpurityReason, futureImpurityReason)) {
+                            dependencies.add(statement);
+                        }
+                    }
+                }
+            }
+        }
+        return dependencies;
+    }
+
+    private hasImpurityReasonEffectOnFuture(reasonPast: ImpurityReason, reasonFuture: ImpurityReason) {
+        // Writes have an effect on reads, if the file is the same
+        if (reasonPast instanceof FileWrite && reasonFuture instanceof FileRead) {
+            return reasonPast.path === reasonFuture.path;
+        }
+        // Writes have an effect on other writes, if the file is the same
+        if (reasonPast instanceof FileWrite && reasonFuture instanceof FileWrite) {
+            return reasonPast.path === reasonFuture.path;
+        }
+        // Write only have effects on reads and writes
+        if (
+            reasonPast instanceof FileWrite &&
+            !(reasonFuture instanceof FileRead) &&
+            !(reasonFuture instanceof FileWrite)
+        ) {
+            return false;
+        }
+        // Reads are only affected by writes
+        if (!(reasonPast instanceof FileWrite) && reasonFuture instanceof FileRead) {
+            return false;
+        }
+        // Other reasons can't affect writes
+        if (reasonFuture instanceof FileWrite) {
+            return false;
+        }
+        // Reads can't affect other reasons
+        if (reasonPast instanceof FileRead) {
+            return false;
+        }
+        // Endless recursions don't have any effect on others
+        if (reasonPast === EndlessRecursion || reasonFuture === EndlessRecursion) {
+            /* c8 ignore next 2 */
+            return false;
+        }
+        // For OtherImpurityReason, PotentiallyImpureParameterCall, UnknownCallableCall we can't say.
+        // So it might have an effect on other impure functions
+        return true;
+    }
```

Comment:
We could replace the if/else chain by polymorphism, i.e. add a method to `ImpurityReason` whether it affects another, later impurity reason.